File: 2020/Day1/src/Day1.py
import argparse

def get_expenses(filename):
  expenses = list()
  with open(filename, 'r') as file:
    expenses = [int(x) for x in file.readlines()]

  return expenses

def part1(expenses):
  for index in range(0, len(expenses) - 1):
    expense1 = expenses[index]
    for index2 in range(index + 1, len(expenses)):
      expense2 = expenses[index2]

      if expense1 + expense2 == 2020:
        return expense1 * expense2

  return -1

def part2(expenses):
  for index in range(0, len(expenses) - 2):
    expense1 = expenses[index]
    for index2 in range(index + 1, len(expenses) - 1):
      expense2 = expenses[index2]

      for index3 in range(index2 + 1, len(expenses)):
        expense3 = expenses[index3]

        if expense1 + expense2 + expense3 == 2020:
          return expense1 * expense2 * expense3

  return -1

def main():
  parser = argparse.ArgumentParser()
  parser.add_argument('-f', '--filename', default='../input/sample1.txt')
  parser.add_argument('-p', '--part', choices=[1, 2], default=1, type=int)
  args = parser.parse_args()

  expenses = get_expenses(args.filename)
  checksum = -1
  if args.part == 1:
    checksum = part1(expenses)
  else:
    checksum = part2(expenses)

  if checksum == -1:
    print('No matching expenses found!')
  else:
    print(f'Expense checksum: {checksum}')

if __name__ == "__main__":
  main()

File: 2020/Day1/src/Day1.rb
#!/usr/bin/ruby

require 'optparse'
require 'set'

def find_combination_totaling_target(expenses, n, target)
  return expenses.map(&:to_i).combination(n).find { |combo| combo.reduce(&:+) == target }.reduce(&:*)
end

def part1(expenses, target)
  return find_combination_totaling_target(expenses, 2, target)
end

def part2(expenses, target)
  return find_combination_totaling_target(expenses, 3, target)
end

options = {
  :part => 1,
  :filename => "../input/sample.txt"
}

OptionParser.new do |opts|
  opts.banner = "Usage: Day1.rb [options]"
  opts.on("-p PART", "--part=PART", Integer, "Part 1 or 2")
  opts.on("-f FILENAME", "--filename=FILENAME", String, "Which filename to use?")
end.parse!(into: options)

expenses = IO.readlines(options[:filename])
checksum = -1
target = 2020
if options[:part] == 1
  checksum = part1(expenses, target)
elsif options[:part] == 2
  checksum = part2(expenses, target)
end

if checksum != -1
  puts "Checksum: #{checksum}"
end
